subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheGradlePropertiesIntegrationTest.groovy

```diff
@@ -179,15 +180,53 @@ class ConfigurationCacheGradlePropertiesIntegrationTest extends AbstractConfigur
         configurationCache.assertStateStored()
 
         when:
-        System.clearProperty('fromPropertiesFile')
+        System.clearProperty("$systemProp")
         configurationCacheRun fixture.task
 
         then:
-        outputDoesNotContain "because system property 'fromPropertiesFile' has changed"
+        outputDoesNotContain "because system property '$systemProp' has changed"
         outputContains fixture.expectedOutputBeforeChange
         configurationCache.assertStateLoaded()
 
         cleanup:
-        System.clearProperty('fromPropertiesFile')
+        System.clearProperty("$systemProp")
+    }
+
+    def "passing cli override invalidates cache entry only if property was read at configuration time"() {
+        given:
+        String systemProp = "fromPropertiesFile"
+        def configurationCache = newConfigurationCacheFixture()
+        def fixture = spec.createFixtureFor(this, systemProp)
+        fixture.setup()
+
+        when:
+        fixture.cleanPropertyRun()
+
+        then:
+        fixture.assertAfterFirstRun()
+
+        when:
+        fixture.cleanPropertyRun()
+
+        then:
+        configurationCache.assertStateLoaded()
+
+        when:
+        fixture.overriddenPropertyRun()
+
+        then:
+        fixture.assertAfterOverriddenRun(configurationCache)
+
+        when:
+        fixture.overriddenPropertyRun()
+
+        then:
+        configurationCache.assertStateLoaded()
+
+        cleanup:
+        System.clearProperty("$systemProp")
+
+        where:
+        spec << SystemPropertiesCompositeBuildFixture.specs()
     }
 }
```

Comment:
It would be great to also have a test that covers mutation of system properties at runtime. For example, `settings.gradle` has `System.setProperty("foo", "bar")`, then calls `includeBuild`, and the included build overrides `foo` with its `gradle.properties`. Then the task in the root build prints the value of the property at execution time.